daemon/cmd/fqdn_test.go

```diff
@@ -156,6 +176,87 @@ func (ds *DaemonSuite) BenchmarkFqdnCache(c *C) {
 	extractDNSLookups(endpoints, "0.0.0.0/0", "*", "")
 }
 
+// Benchmark_notifyOnDNSMsg stresses the main callback function for the DNS
+// proxy path, which is called on every DNS request and response.
+func (ds *DaemonFQDNSuite) Benchmark_notifyOnDNSMsg(c *C) {
+	var (
+		nameManager             = ds.d.dnsNameManager
+		ciliumIOSel             = api.FQDNSelector{MatchName: "cilium.io"}
+		ciliumIOSelMatchPattern = api.FQDNSelector{MatchPattern: "*cilium.io."}
+		ebpfIOSel               = api.FQDNSelector{MatchName: "ebpf.io"}
+		ciliumDNSRecord         = map[string]*fqdn.DNSIPRecords{
+			dns.FQDN("cilium.io"): {TTL: 60, IPs: []net.IP{net.ParseIP("192.0.2.3")}},
+		}
+		ebpfDNSRecord = map[string]*fqdn.DNSIPRecords{
+			dns.FQDN("ebpf.io"): {TTL: 60, IPs: []net.IP{net.ParseIP("192.0.2.4")}},
+		}
+
+		wg sync.WaitGroup
+	)
+
+	// Register rules (simulates applied policies).
+	selectorsToAdd := api.FQDNSelectorSlice{ciliumIOSel, ciliumIOSelMatchPattern, ebpfIOSel}
+	nameManager.Lock()
+	for _, sel := range selectorsToAdd {
+		nameManager.RegisterForIdentityUpdatesLocked(sel)
+	}
+	nameManager.Unlock()
+
+	// Initialize the endpoints.
+	endpoints := make([]*endpoint.Endpoint, c.N)
+	for i := range endpoints {
+		endpoints[i] = &endpoint.Endpoint{
+			ID:   uint16(c.N % 65000),
+			IPv4: netip.MustParseAddr(fmt.Sprintf("10.96.%d.%d", c.N%8, c.N%256)),
+			SecurityIdentity: &identity.Identity{
+				ID: identity.NumericIdentity(c.N % int(identity.MaximumAllocationIdentity)),
+			},
+			DNSZombies: &fqdn.DNSZombieMappings{
+				Mutex: lock.Mutex{},
+			},
+		}
+		ep := endpoints[i]
+		ep.UpdateLogger(nil)
+		ep.DNSHistory = fqdn.NewDNSCache(0)
+	}
+
+	c.ResetTimer()
+	// Simulate parallel DNS responses from the upstream DNS for cilium.io and
+	// ebpf.io, done by every endpoint.
+	for i := 0; i < c.N; i++ {
+		go func(ep *endpoint.Endpoint) {
+			wg.Add(1)
+			defer wg.Done()
+			c.Assert(ds.d.notifyOnDNSMsg(time.Now(), ep, "10.96.64.8:12345", 0, "10.96.64.1:53", &miekgdns.Msg{
```

Comment:
not sure whether it matters or not but this `epIPPort` parameter seems to be hardcoded to one particular IP for all tests rather than being unique for each IP. I don't know if `10.96.64.8` is guaranteed to be set up in the endpoints loop above since `c.N` isn't predetermined but I wonder if this was intended to select the 'current' endpoint's IP rather than a fixed one..?

(This may be more relevant if my previous comment leads to changes in the benchmark).

Reply:
In the end, this parameter is only used in logging. I decided to hardcode it so we don't spend any time convert `net.IP` to a string in the benchmark. I don't think it has any real impact. I will add a comment with that assumption in the code, so it's clear why to readers.